cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java

```diff
@@ -413,11 +419,12 @@ protected Iterable<SparkCollectionRelationalEngine> getRelationalEngines(StageSp
       return super.getRelationalEngines(stageSpec, stageData);
     }
 
+    //TODO : Decide on this
     // If this stage is not pushed down and it's not in the included stages, we can skip relational transformation
     // in the SQL engine.
-    if (!(stageData instanceof SQLBackedCollection) && !shouldForcePushToSQLEngine(stageSpec.getName())) {
-      return super.getRelationalEngines(stageSpec, stageData);
-    }
+//    if (!(stageData instanceof SQLBackedCollection) && !shouldForcePushToSQLEngine(stageSpec.getName())) {
```

Comment:
I think we need 2 pieces of logic here:

1. Check if the SQLBackedCollection is compatible with the engine. For example, a BigQuery SQLBackedCollection is not compatible with Spark, so this needs to be executed in Spark.
2. If the stage is not pushed down, we should push down, but only when the "cost" of the push operation is zero.

I see a couple things to consider here

1. We need to ensure that the `SQLBackedCollection` is compatible with the engines we return. In general, if the collection is either a Spark of a BigQuery collection, the only engine that can handle it is the corresponding engine.


2. If the engine is Spark, the cost to push/pull is pretty much zero. We can add a new property in the SQL Engine to check if the engine is "local". When the engine is local, we assume that the cost of push/pull is zero, and we can push/pull with little/no cost.

```
// Add to SQLEngine interface
default boolean isLocalEngine() {
  return false;
}
```

If this property is true, and the state is not pushed before the relational transform, we can "push" (for little/no cost) and perform the relational transform operation.